src/awkward1/behaviors/categorical.py

```diff
@@ -110,11 +110,109 @@ def _categorical_equal(one, two):
 awkward1.behavior[awkward1.nplike.numpy.equal, "categorical", "categorical"] = _categorical_equal
 
 
+def is_categorical(array):
+    """
+    Args:
+        array: A possibly-categorical Awkward Array.
+
+    If the `array` is categorical (contains #ak.layout.IndexedArray or
+    #ak.layout.IndexedOptionArray labeled with parameter
+    `"__array__" = "categorical"`), then this function returns True;
+    otherwise, it returns False.
+
+    See also #ak.categories, #ak.to_categorical, #ak.from_categorical.
+    """
+
+    layout = awkward1.operations.convert.to_layout(
+        array, allow_record=False, allow_other=False
+    )
+    return layout.purelist_parameter("__array__") == "categorical"
+
+
+def categories(array, highlevel=True):
+    """
+    Args:
+        array: A possibly-categorical Awkward Array.
+        highlevel (bool): If True, return an #ak.Array; otherwise, return
+            a low-level #ak.layout.Content subclass.
+
+    If the `array` is categorical (contains #ak.layout.IndexedArray or
+    #ak.layout.IndexedOptionArray labeled with parameter
+    `"__array__" = "categorical"`), then this function returns its categories.
+
+    See also #ak.is_categorical, #ak.to_categorical, #ak.from_categorical.
+    """
+
+    output = [None]
+
+    def getfunction(layout, depth):
+        if layout.parameter("__array__") == "categorical":
+            output[0] = layout.content
+            return lambda: layout
+
+        else:
+            return None
+
+    awkward1._util.recursively_apply(
+        awkward1.operations.convert.to_layout(
+            array, allow_record=False, allow_other=False
+        ),
+        getfunction,
+    )
+
+    if output[0] is None:
+        return None
+    elif highlevel:
+        return awkward1._util.wrap(output[0], awkward1._util.behaviorof(array))
+    else:
+        return out
```

Comment:
```suggestion
        return output[0]
```

Reply:
Ouch! I added this directly to master because it is absolutely right.